gammapy/spectrum/reflected.py

```diff
@@ -414,3 +419,122 @@ def plot(self, fig=None, ax=None, cmap=None, idx=None, add_legend=False):
             ax.legend(handles=handles)
 
         return fig, ax, cbar
+
+
+class ReflectedRegionsBackgroundMaker:
+    """Reflected regions background maker.
+
+    Parameters
+    ----------
+    region: `~regions.SkyRegion`
+        On region to compute off regions for.
+    angle_increment : `~astropy.coordinates.Angle`, optional
+        Rotation angle applied when a region falls in an excluded region.
+    min_distance : `~astropy.coordinates.Angle`, optional
+        Minimal distance between two consecutive reflected regions
+    min_distance_input : `~astropy.coordinates.Angle`, optional
+        Minimal distance from input region
+    max_region_number : int, optional
+        Maximum number of regions to use
+    exclusion_mask : `~gammapy.maps.WcsNDMap`, optional
+        Exclusion mask
+    binsz : `~astropy.coordinates.Angle`
+        Bin size of the reference map used for region finding.
+    """
+
+    def __init__(
+        self,
+        region,
+        angle_increment="0.1 rad",
+        min_distance="0 rad",
+        min_distance_input="0.1 rad",
+        max_region_number=10000,
+        exclusion_mask=None,
+        binsz="0.01 deg",
+    ):
+        self.region = region
+        self.binsz = binsz
+        self.exclusion_mask = exclusion_mask
+        self.angle_increment = Angle(angle_increment)
+        self.min_distance = Angle(min_distance)
+        self.min_distance_input = Angle(min_distance_input)
+        self.max_region_number = max_region_number
+
+    def _get_finder(self, observation):
+        return ReflectedRegionsFinder(
+            binsz=self.binsz,
+            exclusion_mask=self.exclusion_mask,
+            center=observation.pointing_radec,
+            region=self.region,
+            min_distance=self.min_distance,
+            min_distance_input=self.min_distance_input,
+            max_region_number=self.max_region_number,
+            angle_increment=self.angle_increment,
+        )
+
+    def make_counts_off(self, dataset, observation):
+        """Make off counts.
+
+        Parameters
+        ----------
+        dataset : `SpectrumDataset`
+            Spectrum dataset.
+        observation : `DatastoreObservation`
+            Data store observation.
+
+
+        Returns
+        -------
+        counts_off : `CountsSpectrum`
+            Off counts.
+        """
+        finder = self._get_finder(observation)
+
+        finder.run()
+        regions = finder.reflected_regions
+
+        region_union = regions[0]
+
+        for region in regions[1:]:
+            region_union = region_union.union(region)
+
+        wcs = finder.reference_map.geom.wcs
+        events_off = observation.events.select_region(region_union, wcs)
+
+        counts_off = dataset.counts.copy()
```

Comment:
This should copy the contents as well. It should be created from scratch to have an empty vector

Reply:
Ouch...that's indeed a serious bug. I guess I'd have noticed, when moving the existing tests over, but it's good to fix it now. Thanks!

Reply:
No, actually I realised `CountsSpectrum.fill_events()` overwrites the existing data (see [here](https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/core.py#L131)). This makes the behaviour different from `Map.fill_by_coord()` which adds the data. I guess we should make it consistent...

Reply:
I changed to not copy and will do the consistency change in a follow up PR (it might affect other parts in `gammapy.spectrum`)...